src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3812,6 +3812,25 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                 tmp = make_TupleItem_t(al, loc, value, index,
                                        ASR::down_cast<ASR::Tuple_t>(type)->m_type[i], nullptr);
                 return false;
+            } else if (ASR::is_a<ASR::Array_t>(*type)) {
+                index = ASRUtils::EXPR(tmp);
+                ASR::expr_t* val = ASRUtils::expr_value(index);
+                if (val && ASR::is_a<ASR::IntegerConstant_t>(*val)) {
+                    if (ASR::down_cast<ASR::IntegerConstant_t>(val)->m_n < 0) {
+                        ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(
+                                                        al, loc, 4));
+                        ASR::expr_t *neg_idx = ASRUtils::expr_value(index);
+                        ASR::expr_t *dim_size;
+                        if (ASRUtils::extract_physical_type(type) != ASR::array_physical_typeType::DescriptorArray)
+                            dim_size = ASR::down_cast<ASR::Array_t>(type)->m_dims[idx].m_length;
+                        else {
+                            ASR::expr_t *idx_expr = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, loc, idx + 1, int_type));
+                            dim_size = ASRUtils::EXPR(ASRUtils::make_ArraySize_t_util(al, loc, value, idx_expr, int_type, nullptr, false));
+                        }
```

Comment:
I think we need not check for descriptor array here. The backend implementation of ArraySize would do this check and appropriately return back the dimension size.

Reply:
But then, is the negative indexing still done at compile time for the fixed-size array? My understanding was that ArraySize would create a new alloca in the backend and couldn't be reduced to a constant at compile time.

Reply:
>But then, is the negative indexing still done at compile time for the fixed-size array?

It would be computed at run-time.

If you want it to work using the compile-time value, then you can provide the compile time value in the `m_value` parameter of `make_ArraySize_t_util()`. For example, you can do:
```console
compile_time_dim_size = ASR::down_cast<ASR::Array_t>(type)->m_dims[idx].m_length;
ASR::expr_t *idx_expr = ASRUtils::EXPR(ASR::make_IntegerConstant_t(al, loc, idx + 1, int_type));
dim_size = ASRUtils::EXPR(ASRUtils::make_ArraySize_t_util(al, loc, value, idx_expr, int_type, compile_time_dim_size, false));
```
If the compile-time value is availabe (for example in the case of a fixed size array), it would be used. If the compile-time value is not available, then the dimension size would be computed at run-time.

Reply:
Got it! Thank you. I will update it in another PR.